### PR TITLE
Trainer multi label

### PR DESCRIPTION
**Pull request by sgugger**
This is a follow-up from #7126. The same kinds of models that can output multiple predictions expect multiple labels (not named "labels") so the evaluation code needs to be changed for this. To support models built by users, I added a `label_names` field in the `TrainingArguments` which contain the label names. It then defaults to `["labels"]` for most models, `["start_positions", "end_positions"]` for question answering models if the user does not set it to work seamlessly for all Transformers models.

I ended up writing a few util functions that concat/numpify for tensors or nested lists/tuples of tensors to avoid testing everywhere in `Trainer`, I think the design is cleaner this way and it also supports model with crazy outputs (if we set `output_attentions=True` for instance). I also added a test for the multiple labels predictions.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1382,8 +1390,7 @@ def prediction_step(
             Tuple[Optional[float], Optional[torch.Tensor], Optional[torch.Tensor]]:
             A tuple with the loss, logits and labels (each being optional).
         """
-        has_labels = any(inputs.get(k) is not None for k in ["labels", "lm_labels", "masked_lm_labels"])
```
Comment (sgugger) [inline src/transformers/trainer.py:1385]: No need to test for old deprecated argument names since they have all been changed in the lib and the user can now set their own name if they have an old model they are still using.

Comment (julien-c) [reply src/transformers/trainer.py:1385]: good! this will reduce cruft

File: src/transformers/trainer_utils.py [hunk 2]
```diff
@@ -132,9 +132,49 @@ class HPSearchBackend(ExplicitEnum):
...
+def nested_xla_mesh_reduce(tensors, name):
+    if is_torch_tpu_available():
+        import torch_xla.core.xla_model as xm
+
+        if isinstance(tensors, (list, tuple)):
+            return type(tensors)(nested_xla_mesh_reduce(t, f"{name}_{i}") for i, t in enumerate(tensors))
+        return xm.mesh_reduce(name, tensors, torch.cat)
+    else:
+        raise ImportError("Torch xla must be installed to use `nested_xla_mesh_reduce`")
+
```
Comment (LysandreJik) [inline src/transformers/trainer_utils.py:171]: Did you get a chance to test this on TPU?

**Review by LysandreJik — APPROVED** [review]
LGTM, very nice addition.

File: src/transformers/trainer_utils.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/trainer_utils.py:171]: No, was planning to ask you about it this morning.
